text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+of a parameter).
+
+Finally, *system* libraries and other shared libraries typically use the C ABI
+as well. Software making a Linux `.so`, Windows DLL, or macOS `dylib`, will
+typically expose a C-compatible ABI, and cannot easily provide a higher-level
+safe ABI without shipping language-specific high-level bindings.
+
+crABI defines a common way to make calls across high-level languages, passing
+high-level data types, without dropping to the lowest common denominator of C.
+crABI will work with any language providing a C-compatible FFI (including C
+itself), and languages can also add specific higher-level native support for
+crABI.
+
+crABI aims to be a reasonable default for compiled libraries in both static and
+dynamic form, including system libraries.
+
+This proposal provides version 1 of crABI, with basic support for various
+common data types. Future versions will support additional data types and
+functionality.
+
+# C-based ABI
+[c-based-abi]: #c-based-abi
+
+crABI uses the target's C ABI as a base. For example, on x86-64 Linux, crABI
+uses the [x86-64 psABI](https://gitlab.com/x86-psABIs/x86-64-ABI) as a base,
+just like `extern "C"` does.
+
+Any type whose ABI is already defined by C will be passed through crABI
+identically. Types defined by crABI that the C ABI does not support will be
+translated into a representation using types the C ABI supports (potentially
+indirectly via other crABI-supported types).
+
+crABI does not extend the target C ABI in any ways that would require the use
+of target-specific assembly language to support, such as the use of CPU flags
+or registers not already defined by the C ABI. Doing so would require specific
+enablement work in other languages or runtimes, rather than just building on
+existing support for C FFI.
+
+# Rust language interface and interaction with other features
+[language-interface]: #language-interface
+
+The Rust implementation of crABI will include:
+- An ABI for defining functions or referencing external functions:
+  `extern "crabi"`.
+- A repr for laying out data structures (`struct`, `union`, `enum`) compatible
+  with crABI: `repr(crabi)`.
+- A new lint `improper_crabi_types`, analogous to `improper_ctypes`. This lint
+  defaults to `deny`.
+
+crABI structures and functions also support any structure defined using
+`repr(C)`, as well as enums that have a discriminant type specified.
+
+crABI structures and functions support types defined using `repr(transparent)`
+if they would support the type of the field.
+
+For an enum, the crABI `repr` allows additionally specifying the discriminant
+type: `repr(crabi, u8)`. If an `enum` specifies `repr(crabi)` but does not
+specify a discriminant type, the enum is guaranteed to use the smallest
+discriminant type that holds the maximum discriminant value used by a variant
+in the enum. (This differs from the behavior of `repr(C)` enums without a
+discriminant type.)
```

Comment:
I don't think this should differ. For a start it contradicts the paragraph about C ABI compatibility at lines 69-72 above. The representation of an `enum` in C is implementation-defined as far as the standard is concerned, but in practice you can assume it is a well-defined function, for each platform, of the minimum and maximum enumeration values. (Typically, it has the same representation as the smallest standard integer type that can represent the elements potentially subject to some minimum width.)

I don't think users will expect `repr(crabi)` to differ from `repr(C)` here. It is better to write `repr(crabi, u8)` if that's what you mean.

What we need to specify is what the niche value is for `Option` of an `enum` type, if we are supporting niche-value optimization for `enums` (and I think we should). The obvious answer is "0 if possible, otherwise the maximum value of the chosen integer type if possible, otherwise niche-value optimization is not supported".